src/box/box.cc

```diff
@@ -4412,6 +4412,77 @@ box_process_auth(struct auth_request *request,
 	return 0;
 }
 
+/**
+ * Replica's connection guard:
+ * 1. Ensures that the replica has only one connection at a time.
+ * 2. Calls the replica's disconnection callback in destructor.
+ */
+struct ReplicaConnectionGuard {
+	ReplicaConnectionGuard(struct replica *replica)
+	{
+		replica_ = replica;
+		if (replica == NULL)
+			return;
+		if (replica->is_connected) {
+			tnt_raise(ClientError, ER_CFG, "replication",
+				  "duplicate connection with the same replica "
+				  "UUID");
+		}
+		replica->is_connected = true;
+	}
+	~ReplicaConnectionGuard()
+	{
+		if (replica_ == NULL)
+			return;
+		replica_->is_connected = false;
+		replica_on_disconnect(replica_);
+	}
+	ReplicaConnectionGuard(const ReplicaConnectionGuard &other) = delete;
+	ReplicaConnectionGuard(ReplicaConnectionGuard &&other)
+	{
+		replica_ = other.replica_;
+		other.replica_ = NULL;
+	}
+private:
+	/** Connected replica. */
+	struct replica *replica_;
+};
+
+/**
+ * A helper for replication endpoints to handle a connecting replica.
+ * 1. Checks if the replica doesn't have another connection.
+ * 2. Creates an anonymous replica object if it doesn't exist.
+ * 3. Handles WAL GC state of the replica. It's important that if the replica
+ *    already has a WAL GC consumer, it is re-created and never will be deleted
+ *    so the replica won't lose its consumer, even in the case of an error.
+ * 4. Returns a guard that protects against duplicate replica connection.
+ */
+NODISCARD static ReplicaConnectionGuard
+box_connect_replica(const struct tt_uuid *uuid, const struct vclock *gc_vclock,
+		    struct replica **out)
+{
+	if (tt_uuid_is_nil(uuid)) {
+		*out = NULL;
+		return ReplicaConnectionGuard(NULL);
+	}
+
+	struct replica *replica = replica_by_uuid(uuid);
+	if (replica == NULL)
+		replica = replicaset_add_anon(uuid);
```

Comment:
Well-well, someone hasn't read the commit message 😄 

```console
Important to note that the commit changes the structure of REGISTER and
JOIN endpoints. Before the commit they used the following steps:
1. Retrieve the replica object and clear its WAL GC state, if any.
2. Register a WAL GC consumer but not put it to the replica object.
3. After successful registration, put the created consumer to the replica.
The problem here is if the registration fails, replica loses its
in-memory consumer, so we will never reach if from anonymous replica
garbage collector and its persistent WAL GC consumer will linger in the
system space. To solve this problem, let's register the new WAL GC consumer
and put it to the replica right after deleting the old one. If there is no
corresponding replica object in memory, let's create an anonymous one.
So actual behavior change is that any replica is actually registered as
an anonymous one before the registration. The good thing is we are going
to implement a garbage collector for disconnected anonymous replicas, so
failed to register replicas will be deleted eventually.
```

Also, the function's description (see pt.3):
```c++
/**
 * A helper for replication endpoints to handle a connecting replica.
 * 1. Checks if the replica doesn't have another connection.
 * 2. Creates an anonymous replica object if it doesn't exist.
 * 3. Handles WAL GC state of the replica. It's important that if the replica
 *    already has a WAL GC consumer, it is re-created and never will be deleted
 *    so the replica won't lose its consumer, even in the case of an error.
 * 4. Returns a guard that protects against duplicate replica connection.
 */
NODISCARD static ReplicaConnectionGuard
box_connect_replica(const struct tt_uuid *uuid, const struct vclock *gc_vclock,
		    struct replica **out);
```

Reply:
Funny how we fallback to C++ sometimes, accepting its superiority in cases like this, and how much do we need in C to mimic even just trivial destructors. But as soon as there is a suggestion to completely switch to C++, some people would cry from phantom pains of imaginary performance drops 🤔. @sergepetrenko, just the simplest C++ usage as "C with classes" would allow to drop so much code ...

Reply:
Have you ever seen how we [emulate](https://github.com/tarantool/tarantool/blob/master/src/lib/salad/bps_tree.h) C++ templates in B-tree? Brings me to tears faster than any Ernest Hemingway story.